superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/index.tsx

```diff
@@ -131,13 +135,14 @@ const groupByControl: SharedControlConfig<'SelectControl', ColumnMeta> = {
   promptTextCreator: (label: unknown) => label,
   mapStateToProps(state, { includeTime }) {
     const newState: ExtraControlProps = {};
-    if (state.datasource) {
-      const options = state.datasource.columns.filter(c => c.groupby);
+    const { datasource } = state;
+    if (datasource?.columns[0]?.hasOwnProperty('groupby')) {
+      const options = (datasource as Dataset).columns.filter(c => c.groupby);
       if (includeTime) {
-        options.unshift(TIME_COLUMN_OPTION);
+        options.unshift(DATASET_TIME_COLUMN_OPTION);
       }
       newState.options = options;
-    }
+    } else newState.options = datasource?.columns;
```

Comment:
we should still do lines 141 and 142 if this is a query.

Reply:
Oh good catch, thank you! Fixed in [`this commit`](https://github.com/apache/superset/pull/19855/commits/faac3911412d0a9371261ff1ed3f207cfc9d741b).